Various ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr cleanups

kmiller68 opened the pull request:
#### cb9e03866203faa7c3a986f3bbd561661cca913f
<pre>
Various ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr cleanups
<a href="https://bugs.webkit.org/show_bug.cgi?id=283588">https://bugs.webkit.org/show_bug.cgi?id=283588</a>
<a href="https://rdar.apple.com/140325747">rdar://140325747</a>

Reviewed by Alex Christensen.

This patch clears up some parts of how ThreadSafeRefCountedAndCanMakeWeakPtr work.
In particular, now that we no longer permit creating a WeakRef during the destructor
of an object we don&apos;t need to use a HashMap in ThreadSafeWeakHashSet and can just use
a std::pair, which simplifies the logic. I also added an optimization to not check the
set if there&apos;s no weakRefCount when removing/contains checking. This also prevents
allocating a ControlBlock for an object that can&apos;t be in the table anyway.

I was originally thinking we could just remove this second pointer but due to
multiple inheritance it&apos;s actually necessary. I added some comments to clarify this.

I also added some logic to remove an entry if it&apos;s stale when calling
`ThreadSafeWeakHashSet::contains()` as well as some other places we could
opportunistically clear dead entries.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/AbstractThreadSafeRefCountedAndCanMakeWeakPtr.h: Added.
* Source/WTF/wtf/ThreadSafeWeakHashSet.h:
* Source/WTF/wtf/ThreadSafeWeakPtr.h:
(WTF::ThreadSafeWeakPtrControlBlock::weakRef):
(WTF::ThreadSafeWeakPtrControlBlock::makeStrongReferenceIfPossible const):
(WTF::ThreadSafeWeakPtrControlBlock::weakRefCount const):
(WTF::ThreadSafeWeakPtrControlBlock::refCount const):
(WTF::ThreadSafeWeakPtrControlBlock::hasOneRef const):
(WTF::ThreadSafeWeakPtrControlBlock::ThreadSafeWeakPtrControlBlock):
(WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::controlBlock const):
(WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::weakRefCount const):
(WTF::ThreadSafeWeakPtrControlBlock::weakReferenceCount const): Deleted.
* Source/WebCore/Modules/mediastream/CanvasCaptureMediaStreamTrack.h:
* Source/WebCore/Modules/mediastream/VideoTrackGenerator.h:
* Source/WebCore/Modules/webaudio/MediaStreamAudioSource.h:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::sinkIntoSerializedImageBuffer):
* Source/WebCore/platform/mediastream/RealtimeIncomingAudioSource.h:
* Source/WebCore/platform/mediastream/RealtimeIncomingVideoSource.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
(WebCore::RealtimeVideoCaptureSource::ref const): Deleted.
(WebCore::RealtimeVideoCaptureSource::deref const): Deleted.
(WebCore::RealtimeVideoCaptureSource::controlBlock const): Deleted.
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCaptureSource.h:
* Source/WebCore/platform/mediastream/gstreamer/RealtimeIncomingSourceGStreamer.h:
(): Deleted.
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.h:
* Source/WebCore/platform/mock/MockRealtimeAudioSource.h:
* Source/WebCore/platform/xr/PlatformXR.h:
(PlatformXR::Device::ref const): Deleted.
(PlatformXR::Device::deref const): Deleted.
(PlatformXR::Device::controlBlock const): Deleted.
* Source/WebKit/UIProcess/SpeechRecognitionRemoteRealtimeMediaSource.h:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
* Source/WebKit/WebProcess/GPU/graphics/wc/RemoteWCLayerTreeHostProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioHardwareListener.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.h:
* Source/WebKit/WebProcess/GPU/media/RemoteImageDecoderAVFManager.h:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.h:
(): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteRemoteCommandListener.h:
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.h:
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.h:
* Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxyProcessor.h:
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessConnection.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.h:
(): Deleted.
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.h:
* Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
(TestWebKitAPI::ObjectAddingAndRemovingItself::~ObjectAddingAndRemovingItself):
(TestWebKitAPI::ObjectRemovingItself::create):
(TestWebKitAPI::ObjectRemovingItself::~ObjectRemovingItself):
(TestWebKitAPI::ObjectRemovingItself::ObjectRemovingItself):
(TestWebKitAPI::TEST(WTF_ThreadSafeWeakPtr, ThreadSafeWeakHashSetRemoveNonExistantOnDestruction)):
(TestWebKitAPI::TEST(WTF_ThreadSafeWeakPtr, MultipleInheritance)):

Canonical link: <a href="https://commits.webkit.org/287313@main">https://commits.webkit.org/287313@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/66accc562e628fdf6d9c2e5a9f2c80330bc099e2

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/79086 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/58122 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/32463 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/83725 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/30299 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/81219 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/67230 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/6392 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/61899 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19814 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/82153 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51939 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/72084 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/42203 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/49290 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/26193 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/28664 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/72210 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/70406 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/26612 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/85112 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/78303 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/6409 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/4432 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/70138 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/6572 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67955 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/69387 "Passed tests") | 
| [  ~~🛠 🧪 merge~~](https://ews-build.webkit.org/#/builders/19/builds/17300 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/13432 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/12224 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/100626 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/6361 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/12181 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/21960 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/6317 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/9772 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/8109 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

MenloDorian reviewed (commented):
In the commit message, for clarity, please use ThreadSafeRefCountedAndCanMakeWeakPtr instead of TSRFACMWP.  BTW, it would have been TSRCACMWP, not TSRFACMWP.

MenloDorian commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -92,55 +92,64 @@ class ThreadSafeWeakHashSet final {
-        auto it = m_map.find(static_cast<const T*>(&value));
-        if (it == m_map.end())
+        // FIXME: If there's no weak refs then it can't be in our table. In that case
+        // there's no point in potentially allocating a ControlBlock.
```
Please add a rdar link here to track the work for the FIXME in the future.

MenloDorian commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -92,55 +92,64 @@ class ThreadSafeWeakHashSet final {
-        auto it = m_map.find(static_cast<const T*>(&value));
-        if (it == m_map.end())
+        // FIXME: If there's no weak refs then it can't be in our table. In that case
+        // there's no point in potentially allocating a ControlBlock.
```
Please add the rdar link here too to track the FIXME work in the future.

MenloDorian commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -92,55 +92,64 @@ class ThreadSafeWeakHashSet final {
         amortizedCleanupIfNeeded();
-        for (auto& controlBlock : m_map.values()) {
-            if (!controlBlock->objectHasStartedDeletion())
+        // FIXME: This seems like it should remove any stale entries it finds.
```
Please add a radar link for this FIXME as well.

MenloDorian commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -92,55 +92,64 @@ class ThreadSafeWeakHashSet final {
-            if (!controlBlock->objectHasStartedDeletion())
+        // FIXME: This seems like it should remove any stale entries it finds.
+        for (auto& pair : m_set) {
+            if (!pair.first->objectHasStartedDeletion())
```
Does this also need to check if `pair.first` as we do in the `remove` and `contains` implementations above?

kmiller68 replied:
Actually I think none of the `!.*first` checks are needed since we don't put null ControlBlocks into the set.

MenloDorian commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -152,14 +161,14 @@ class ThreadSafeWeakHashSet final {
-            strongReferences = compactMap(m_map, [&hasNullReferences](auto& pair) -> RefPtr<T> {
-                if (RefPtr strongReference = pair.value->template makeStrongReferenceIfPossible<T>(pair.key))
+            strongReferences = compactMap(m_set, [&hasNullReferences](auto& pair) -> RefPtr<T> {
+                if (RefPtr strongReference = pair.first->template makeStrongReferenceIfPossible<T>(pair.second))
```
Ditto: does this need a `pair.first` null check too?

MenloDorian commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -152,14 +161,14 @@ class ThreadSafeWeakHashSet final {
             });
             if (hasNullReferences)
-                m_map.removeIf([](auto& pair) { return pair.value->objectHasStartedDeletion(); });
+                m_set.removeIf([](auto& pair) { return pair.first->objectHasStartedDeletion(); });
```
Ditto: need null check?

MenloDorian commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -187,36 +197,37 @@ class ThreadSafeWeakHashSet final {
-            m_map.removeIf([] (auto& pair) {
-                return pair.value->objectHasStartedDeletion();
+            m_set.removeIf([] (auto& pair) {
+                ASSERT(pair.first->weakRefCount());
+                return pair.first->objectHasStartedDeletion();
```
Same here.  Do we need to null check `pair.first`?

MenloDorian commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -122,29 +94,74 @@ class ThreadSafeWeakPtrControlBlock {
+    // if (!objectHasStartedDeletion())
+    //     strongRef();
+    // Between objectHasStartedDeletion() and strongRef() another thread holding the sole remaining reference
+    // to the underlying object could release it's reference and start deletion.
```
/could release/can release/

kmiller68 replied:
Isn't either valid here? I suppose maybe the connotation is different?

MenloDorian commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -122,29 +94,74 @@ class ThreadSafeWeakPtrControlBlock {
+    // Between objectHasStartedDeletion() and strongRef() another thread holding the sole remaining reference
+    // to the underlying object could release it's reference and start deletion.
+
+    // FIXME: These also shouldn't need to take the lock.
```
Please add a rdar link for this FIXME work in the future too.

MenloDorian commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -422,8 +441,7 @@ class ThreadSafeWeakPtr {
     TaggedPtr<T, TaggingTraits> m_objectOfCorrectType;
-    // FIXME: Either remove ThreadSafeWeakPtrControlBlock::m_object as redundant information,
-    // or use CompactRefPtrTuple to reduce sizeof(ThreadSafeWeakPtr) by storing just an offset
+    // FIXME: Use CompactRefPtrTuple to reduce sizeof(ThreadSafeWeakPtr) by storing just an offset
```
Please add a radar ink here too.  Out maybe there should can use the same 1 radar link for all these FIXMEs.

kmiller68 replied:
Done.

jyavenard commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -92,55 +92,70 @@ class ThreadSafeWeakHashSet final {
-        if (it == m_map.end())
+        // If there's no weak refs then it can't be in our table. In that case
+        // there's no point in potentially allocating a ControlBlock.
+        if (!value.weakRefCount())
```
this seems to be potentially racy if a weakRef is taken between the time it's removed.
The previous code avoided this situation completely.

kmiller68 replied:
@jyavenard  and I had a discussion about this offline. There's not actually a race here since the presence of the value's controlBlock in the set implies it must have a weakRefCount of at least 1 (the set's ref). Thus the contrapositive says a weakRefCount of less than 1 (aka 0) means it can't be in the set.

jyavenard commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -92,55 +92,70 @@ class ThreadSafeWeakHashSet final {
-        if (it == m_map.end())
+        // If there's no weak refs then it can't be in our table. In that case
+        // there's no point in potentially allocating a ControlBlock.
+        if (!value.weakRefCount())
```
same

jyavenard commented on Source/WebCore/Modules/mediastream/CanvasCaptureMediaStreamTrack.h:
```diff
@@ -60,6 +60,7 @@ class CanvasCaptureMediaStreamTrack final : public MediaStreamTrack {
         void ref() const final { ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Source, WTF::DestructionThread::MainRunLoop>::ref(); }
         void deref() const final { ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Source, WTF::DestructionThread::MainRunLoop>::deref(); }
         ThreadSafeWeakPtrControlBlock& controlBlock() const final { return ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Source, WTF::DestructionThread::MainRunLoop>::controlBlock(); }
+        size_t weakRefCount() const final { return ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Source, WTF::DestructionThread::MainRunLoop>::weakRefCount(); }
```
Having to add another virtual method for classes implementing `ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr` seems like a very unfortunate requirement. 

And as weakRefCount isn't an atomic operation, there's no telling it can't change between the time we read the value and the time we perform the optimisation of not looking it up in the table.

achristensen07 commented on Source/WebCore/Modules/webaudio/MediaStreamAudioSource.h:
```diff
@@ -45,7 +45,7 @@ class MediaStreamAudioSource final : public RealtimeMediaSource, public ThreadSa
     void ref() const final { ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<MediaStreamAudioSource, WTF::DestructionThread::MainRunLoop>::ref(); }
     void deref() const final { ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<MediaStreamAudioSource, WTF::DestructionThread::MainRunLoop>::deref(); }
-    ThreadSafeWeakPtrControlBlock& controlBlock() const final { return ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<MediaStreamAudioSource, WTF::DestructionThread::MainRunLoop>::controlBlock(); }
+    ThreadSafeWeakPtrControlBlock& controlBlock() const final { return ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<MediaStreamAudioSource, WTF::DestructionThread::MainRunLoop>::controlBlock(); }    size_t weakRefCount() const final { return ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<MediaStreamAudioSource, WTF::DestructionThread::MainRunLoop>::weakRefCount(); }
```
These shouldn't be on the same line.

kmiller68 replied:
Fixed.

achristensen07 commented on Source/WebCore/platform/xr/PlatformXR.h:
```diff
@@ -368,7 +368,7 @@ class Device : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Device> {
     void ref() const { ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Device>::ref(); }
     void deref() const { ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Device>::deref(); }
-    ThreadSafeWeakPtrControlBlock& controlBlock() const { return ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Device>::controlBlock(); }
+    ThreadSafeWeakPtrControlBlock& controlBlock() const { return ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Device>::controlBlock(); }    size_t weakRefCount() const { return ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Device>::weakRefCount(); }
```
Ditto.  Also, I think the &lt;Device&gt; isn't needed in all these cases.  Could be a nice little cleanup.

kmiller68 replied:
It seems like it shouldn't need to be defined at all since they're not virtual anyway. I removed them.

achristensen07 commented on Source/WTF/wtf/AbstractThreadSafeRefCountedAndCanMakeWeakPtr.h:
```diff
@@ -0,0 +1,52 @@
+
+// Convinience macro that implements AbstractThreadSafeRefCountedAndCanMakeWeakPtr for you.
+#define WTF_ABSTRACT_THREAD_SAFE_REF_COUNTED_AND_CAN_MAKE_WEAK_PTR_IMPL(Class, ... /* DestructionThread */) \
+    void ref() const final { ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<Class, ##__VA_ARGS__>::ref(); } \
```
I really think &lt;Class, ##__VA_ARGS__&gt; is unnecessary, so you don't need any parameters for this macro.  If that doesn't compile for some reason, this is fine.

kmiller68 replied:
Sure, I'll try it.

achristensen07 commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -92,55 +92,70 @@ class ThreadSafeWeakHashSet final {
         amortizedCleanupIfNeeded();
-        auto it = m_map.find(static_cast<const T*>(&value));
-        if (it == m_map.end())
+        // If there's no weak refs then it can't be in our table. In that case
```
If there are no weak refs...
refs is plural.  Same below.

achristensen07 commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -92,55 +92,70 @@ class ThreadSafeWeakHashSet final {
         amortizedCleanupIfNeeded();
-        for (auto& controlBlock : m_map.values()) {
-            if (!controlBlock->objectHasStartedDeletion())
+        // FIXME: This seems like it should remove any stale entries it finds. https://bugs.webkit.org/show_bug.cgi?id=283928
```
I think it should not.  Right now it can return with the first non-deleted object it finds, but if we started pruning we would have to iterate the entire collection every time.

kmiller68 replied:
What I meant is not that it should do extra work but delete stale entries it sees until we find a live entry.

achristensen07 commented on Source/WTF/wtf/ThreadSafeWeakHashSet.h:
```diff
@@ -169,9 +184,10 @@ class ThreadSafeWeakHashSet final {
     {
         Vector<ThreadSafeWeakPtr<T>> weakReferences;
         {
+            // FIXME: It seems like this should prune known dead entries as it goes. https://bugs.webkit.org/show_bug.cgi?id=283928
```
I think this is a reasonable place to prune since we are iterating the entire set anyways.  Keep this.

achristensen07 commented on Source/WTF/wtf/ThreadSafeWeakPtr.h:
```diff
@@ -43,11 +43,8 @@ class ThreadSafeWeakPtrControlBlock {
-            return this;
-        }
-        return nullptr;
+        ++m_weakReferenceCount;
```
ASSERT(m_object) here since we are removing the null check?  Or keep null check?

kmiller68 replied:
I think folks pass around a dead ControlBlock but I can try adding the ASSERT and see if there are crashes.

achristensen07 commented on Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
```diff
@@ -2971,6 +2968,38 @@ TEST(WTF_ThreadSafeWeakPtr, ThreadSafeWeakHashSetRemoveOnDestruction)
+// Test removing an object that was never inserted during its destructor is ok.
+TEST(WTF_ThreadSafeWeakPtr, ThreadSafeWeakHashSetRemoveNonExistantOnDestruction)
+{
+
```
whitespace

kmiller68 replied:
Fixed.

kmiller68 committed:
Various ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr cleanups
https://bugs.webkit.org/show_bug.cgi?id=283588
rdar://140325747

Reviewed by Alex Christensen.

This patch clears up some parts of how ThreadSafeRefCountedAndCanMakeWeakPtr work.
In particular, now that we no longer permit creating a WeakRef during the destructor
of an object we don't need to use a HashMap in ThreadSafeWeakHashSet and can just use
a std::pair, which simplifies the logic. I also added an optimization to not check the
set if there's no weakRefCount when removing/contains checking. This also prevents
allocating a ControlBlock for an object that can't be in the table anyway.

I was originally thinking we could just remove this second pointer but due to
multiple inheritance it's actually necessary. I added some comments to clarify this.

I also added some logic to remove an entry if it's stale when calling
`ThreadSafeWeakHashSet::contains()` as well as some other places we could
opportunistically clear dead entries.

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/AbstractThreadSafeRefCountedAndCanMakeWeakPtr.h: Added.
* Source/WTF/wtf/ThreadSafeWeakHashSet.h:
* Source/WTF/wtf/ThreadSafeWeakPtr.h:
(WTF::ThreadSafeWeakPtrControlBlock::weakRef):
(WTF::ThreadSafeWeakPtrControlBlock::makeStrongReferenceIfPossible const):
(WTF::ThreadSafeWeakPtrControlBlock::weakRefCount const):
(WTF::ThreadSafeWeakPtrControlBlock::refCount const):
(WTF::ThreadSafeWeakPtrControlBlock::hasOneRef const):
(WTF::ThreadSafeWeakPtrControlBlock::ThreadSafeWeakPtrControlBlock):
(WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::controlBlock const):
(WTF::ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr::weakRefCount const):
(WTF::ThreadSafeWeakPtrControlBlock::weakReferenceCount const): Deleted.
* Source/WebCore/Modules/mediastream/CanvasCaptureMediaStreamTrack.h:
* Source/WebCore/Modules/mediastream/VideoTrackGenerator.h:
* Source/WebCore/Modules/webaudio/MediaStreamAudioSource.h:
* Source/WebCore/platform/graphics/ImageBuffer.cpp:
(WebCore::ImageBuffer::sinkIntoSerializedImageBuffer):
* Source/WebCore/platform/mediastream/RealtimeIncomingAudioSource.h:
* Source/WebCore/platform/mediastream/RealtimeIncomingVideoSource.h:
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
(WebCore::RealtimeVideoCaptureSource::ref const): Deleted.
(WebCore::RealtimeVideoCaptureSource::deref const): Deleted.
(WebCore::RealtimeVideoCaptureSource::controlBlock const): Deleted.
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/cocoa/DisplayCaptureSourceCocoa.h:
* Source/WebCore/platform/mediastream/gstreamer/GStreamerAudioCaptureSource.h:
* Source/WebCore/platform/mediastream/gstreamer/RealtimeIncomingSourceGStreamer.h:
(): Deleted.
* Source/WebCore/platform/mediastream/mac/CoreAudioCaptureSource.h:
* Source/WebCore/platform/mock/MockRealtimeAudioSource.h:
* Source/WebCore/platform/xr/PlatformXR.h:
(PlatformXR::Device::ref const): Deleted.
(PlatformXR::Device::deref const): Deleted.
(PlatformXR::Device::controlBlock const): Deleted.
* Source/WebKit/UIProcess/SpeechRecognitionRemoteRealtimeMediaSource.h:
* Source/WebKit/WebProcess/GPU/GPUProcessConnection.h:
* Source/WebKit/WebProcess/GPU/graphics/wc/RemoteWCLayerTreeHostProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioDestinationProxy.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioHardwareListener.h:
* Source/WebKit/WebProcess/GPU/media/RemoteAudioSession.h:
* Source/WebKit/WebProcess/GPU/media/RemoteImageDecoderAVFManager.h:
* Source/WebKit/WebProcess/GPU/media/RemoteMediaPlayerManager.h:
(): Deleted.
* Source/WebKit/WebProcess/GPU/media/RemoteRemoteCommandListener.h:
* Source/WebKit/WebProcess/GPU/media/ios/RemoteMediaSessionHelper.h:
* Source/WebKit/WebProcess/GPU/webrtc/LibWebRTCCodecs.h:
* Source/WebKit/WebProcess/GPU/webrtc/RemoteVideoFrameObjectHeapProxyProcessor.h:
* Source/WebKit/WebProcess/GPU/webrtc/SampleBufferDisplayLayer.h:
* Source/WebKit/WebProcess/Model/ModelProcessConnection.h:
* Source/WebKit/WebProcess/Model/ModelProcessModelPlayerManager.h:
(): Deleted.
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.h:
* Tools/TestWebKitAPI/Tests/WTF/WeakPtr.cpp:
(TestWebKitAPI::ObjectAddingAndRemovingItself::~ObjectAddingAndRemovingItself):
(TestWebKitAPI::ObjectRemovingItself::create):
(TestWebKitAPI::ObjectRemovingItself::~ObjectRemovingItself):
(TestWebKitAPI::ObjectRemovingItself::ObjectRemovingItself):
(TestWebKitAPI::TEST(WTF_ThreadSafeWeakPtr, ThreadSafeWeakHashSetRemoveNonExistantOnDestruction)):
(TestWebKitAPI::TEST(WTF_ThreadSafeWeakPtr, MultipleInheritance)):

Canonical link: https://commits.webkit.org/287313@main
